SNLS_NLSPowell.h

```diff
@@ -0,0 +1,729 @@
+    __snls_hdev__ 
+    double getRes() const { return m_res; }
+
+    // setX can be used to set the initial guess
+    __snls_hdev__ 
+    inline 
+    void setX( const double* const x ) 
+    {
+        for (int iX = 0; iX < m_nDim; ++iX) {
+            m_x[iX] = x[iX] ;
+        }
+    };
+
+    __snls_hdev__ 
+    inline 
+    void getX( double* const x ) const 
+    {
+        for (int iX = 0; iX < m_nDim; ++iX) {
+            x[iX] = m_x[iX] ;
+        }
+    };   
+
+    /**
+     * Must call setupSolver before calling solve
+    */
+    __snls_hdev__ 
+    inline
+    void setupSolver(int maxIter,
+                     double x_tolerance,
+                     double f_tolerance,
+                     double factor = 100.0,
+                     int outputLevel = 0) 
+    {
+        m_status = unConverged;
+        m_fevals = 0;
+        m_jevals = 0;
+        m_factor = factor;
+
+        m_maxIter = maxIter;
+        m_maxfev = maxIter;
+        m_x_tolerance = x_tolerance;
+        m_f_tolerance = f_tolerance;
+
+        this->setOutputlevel( outputLevel );
+
+        m_complete = true;
+
+    }
+
+    __snls_hdev__ 
+    void setOutputlevel( int outputLevel ) {
+        m_outputLevel = outputLevel;
+        m_os          = nullptr;
+        //
+#ifdef __cuda_host_only__
+        if ( m_outputLevel > 0 ) {
+            m_os = &(std::cout);
+        }
+#endif
+    }
+
+    // solve returns status
+    //
+    // on exit, _res is consistent with _x
+    __snls_hdev__ 
+    inline
+    SNLSStatus_t solve() {
+    
+        if ( !m_complete) {
+        SNLS_FAIL("solve", "Setup not called beforehand!") ;
+        }
+
+        double residual[m_nDim];
+        double Jacobian[m_nXnDim];
+
+        // Working arrays
+        double Qmat[m_nXnDim];
+        double work_arr1[m_nDim];
+        double work_arr2[m_nDim];
+        double work_arr3[m_nDim];
+        double qtf[m_nDim];
+        // Only use diag if we need it
+        double diag[m_nDim];
+
+        if (!scaleDiag) {
+            for (int i = 0; i < m_nDim; i++) {
+                diag[i] = 1.0;
+            }
+        }
+
+        m_status = solveInit(residual);
+        if (m_status != unConverged) {
+            return m_status;
+        }
+
+        while (m_status == unConverged && m_nIters < m_maxIter && m_fevals < m_maxIter) {
+            solveStep(residual, Jacobian, Qmat, diag, qtf, work_arr1, work_arr2, work_arr3);
+        }
+
+        return m_status;
+    }
+
+    // convenience wrapper, for the current _x
+    __snls_hdev__ 
+    inline bool computeRJ(double* const r,
+                          double* const J ) 
+    {
+        bool retval = this->m_crj.computeRJ(r, J, m_x);
+        // Add the finite difference portion of things back in here at a later point        
+        return retval;  
+    }
+
+    private:
+
+    __snls_hdev__
+    inline
+    SNLSStatus_t solveInit(double* const residual)
+    {
+        m_jevals = 0;
+        m_fevals = 1; 
```

Comment:
If you are always using this->computeRJ to do function and residual evaluations, then it might make more sense to initialize m_fevals to 0 and put the increment of m_fevals and m_jevals inside of this->computeRJ (with a conditional on the J being null).

Reply:
I've thought of that, and I can't remember why I was avoiding it originally.

Reply:
Made this change, but I still reset them to zero when this called.